vault/router.go

```diff
@@ -174,6 +181,31 @@ func (r *Router) MatchingMount(path string) string {
 	return mount
 }
 
+//MatchingPrefix returns a mount prefix that a path may be a part of
+func (r *Router) MatchingPrefix(path string) string {
```

Comment:
This function doesn't get a read lock but should, as it can be called externally (via MountConflict from the mount code). However, it is called with a lock already held from the router's Mount function.

The way to deal with this would be to have MatchingPrefix grab a read lock, then call matchingPrefixInternal which has the actual logic. The router's Mount function would then just call matchingPrefixInternal directly.